## Machine Learning Pipeline

As the class practice, the students will be required to develop a machine learning pipeline using the `Churn_Modelling_train_test.csv` dataset.

**About dataset**

This dataset is obained from [kaggle](https://www.kaggle.com/datasets/shubhammeshram579/bank-customer-churn-prediction?resource=download). It contains information on bank customers who either left the bank or continue to be a customer. The dataset includes the following attributes:

* Customer ID: A unique identifier for each customer
* Surname: The customer's surname or last name
* Credit Score: A numerical value representing the customer's credit score
* Geography: The country where the customer resides (France, Spain or Germany)
* Gender: The customer's gender (Male or Female)
* Age: The customer's age.
* Tenure: The number of years the customer has been with the bank
* Balance: The customer's account balance
* NumOfProducts: The number of bank products the customer uses (e.g., savings account, credit card)
* HasCrCard: Whether the customer has a credit card (1 = yes, 0 = no)
* IsActiveMember: Whether the customer is an active member (1 = yes, 0 = no)
* EstimatedSalary: The estimated salary of the customer
* Exited: Whether the customer has churned (1 = yes, 0 = no)


### Exploratory Data Analysis

In this section the students are required to do an EDA to understand the dataset.

In [27]:
import mlflow
import mlflow.sklearn

In [28]:
# import packages
# import libraries
from ipykernel.kernelspec import install
import pandas as pd
import joblib
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression   
from sklearn.preprocessing import OneHotEncoder      
from sklearn.metrics import f1_score


In [29]:
# Load the dataset
df = pd.read_csv("/Users/mac/Desktop/Eada Masters/Semester 3/ML operations and System Design /Excercises /mlops-and-system-design/class 2 exercise/Churn_Modelling_train_test.csv")

In [30]:
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,4784,15729224,Jennings,710,France,Female,37.0,5,0.00,2,1.0,0.0,115403.31,0
1,1497,15799156,Okwuadigbo,569,Spain,Male,38.0,8,0.00,2,0.0,0.0,79618.79,0
2,1958,15674922,Beavers,710,France,Male,54.0,6,171137.62,1,1.0,1.0,167023.95,1
3,9174,15653572,Thornton,673,Spain,Male,43.0,8,127132.96,1,0.0,1.0,6009.27,1
4,9748,15775761,Iweobiegbunam,610,Germany,Female,69.0,5,86038.21,3,0.0,0.0,192743.06,1


In [31]:
# dataset analysis
df.info()
df.describe()
df.isna().sum()
df['Age'].mean()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9001 entries, 0 to 9000
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9001 non-null   int64  
 1   CustomerId       9001 non-null   int64  
 2   Surname          9001 non-null   object 
 3   CreditScore      9001 non-null   int64  
 4   Geography        9000 non-null   object 
 5   Gender           9001 non-null   object 
 6   Age              9000 non-null   float64
 7   Tenure           9001 non-null   int64  
 8   Balance          9001 non-null   float64
 9   NumOfProducts    9001 non-null   int64  
 10  HasCrCard        9000 non-null   float64
 11  IsActiveMember   9001 non-null   float64
 12  EstimatedSalary  9001 non-null   float64
 13  Exited           9001 non-null   int64  
dtypes: float64(5), int64(6), object(3)
memory usage: 984.6+ KB


np.float64(38.90178111111111)

In [32]:
# review categorical variables
categorical_cols = ['Geography', 'Gender']
for col in categorical_cols:
    print(f"\n{col} distribution:")
    print(df[col].value_counts())
    print(f"{col} unique values: {df[col].nunique()}")


Geography distribution:
Geography
France     4505
Spain      2250
Germany    2245
Name: count, dtype: int64
Geography unique values: 3

Gender distribution:
Gender
Male      4906
Female    4095
Name: count, dtype: int64
Gender unique values: 2


In [33]:
# review binary variables
binary_cols = ['HasCrCard', 'IsActiveMember', 'Exited']
for col in binary_cols:
    print(f"\n{col} distribution:")
    print(df[col].value_counts())
    print(f"{col} percentage:")
    print(df[col].value_counts(normalize=True) * 100)


HasCrCard distribution:
HasCrCard
1.0    6351
0.0    2649
Name: count, dtype: int64
HasCrCard percentage:
HasCrCard
1.0    70.566667
0.0    29.433333
Name: proportion, dtype: float64

IsActiveMember distribution:
IsActiveMember
1.0    4633
0.0    4368
Name: count, dtype: int64
IsActiveMember percentage:
IsActiveMember
1.0    51.472059
0.0    48.527941
Name: proportion, dtype: float64

Exited distribution:
Exited
0    7161
1    1840
Name: count, dtype: int64
Exited percentage:
Exited
0    79.557827
1    20.442173
Name: proportion, dtype: float64


In [34]:
# review numerical variables
# review numerical variables
...
print("Mean", df['Age'].mean())

print("Median", df['Age'].median())

print("Min", df['Age'].min())

print("Max", df['Age'].max())

print("Std", df['Age'].std())


Mean 38.90178111111111
Median 37.0
Min 18.0
Max 92.0
Std 10.450759502027477


In [35]:
print("Mean", df['Balance'].mean())

print("Median", df['Balance'].median())

print("Min", df['Balance'].min())

print("Max", df['Balance'].max())

print("Std", df['Balance'].std())

Mean 76222.21082657482
Median 96997.09
Min 0.0
Max 250898.09
Std 62432.19815106162


In [36]:
print("Mean", df['Tenure'].mean())

print("Median", df['Tenure'].median())

print("Min", df['Tenure'].min())

print("Max", df['Tenure'].max())

print("Std", df['Tenure'].std())

Mean 5.007888012443062
Median 5.0
Min 0
Max 10
Std 2.8940252619475744


In [37]:
print("Mean", df['NumOfProducts'].mean())

print("Median", df['NumOfProducts'].median())

print("Min", df['NumOfProducts'].min())

print("Max", df['NumOfProducts'].max())

print("Std", df['NumOfProducts'].std())

Mean 1.5314965003888457
Median 1.0
Min 1
Max 4
Std 0.5793982263367707


In [38]:
print("Mean", df['EstimatedSalary'].mean())

print("Median", df['EstimatedSalary'].median())

print("Min", df['EstimatedSalary'].min())

print("Max", df['EstimatedSalary'].max())

print("Std", df['EstimatedSalary'].std())

Mean 100180.82396733695
Median 100556.98
Min 11.58
Max 199992.48
Std 57561.18953385744


In [39]:
print("Mean", df['CreditScore'].mean())

print("Median", df['CreditScore'].median())

print("Min", df['CreditScore'].min())

print("Max", df['CreditScore'].max())

print("Std", df['CreditScore'].std())

Mean 650.681368736807
Median 652.0
Min 350
Max 850
Std 96.53959128284978


In [40]:
cols = [
    'Balance',
    'CreditScore',
    'Age',
    'Tenure',
    'NumOfProducts',
    'EstimatedSalary'
]

for col in cols:

    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)

    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = df[(df[col] < lower) | (df[col] > upper)]

    print(f"{col}: {outliers.shape[0]} outliers")

Balance: 0 outliers
CreditScore: 13 outliers
Age: 318 outliers
Tenure: 0 outliers
NumOfProducts: 50 outliers
EstimatedSalary: 0 outliers


**EDA Conclusions**

From this Exploratory Data Analysis we can retrieve the following information:

Most of the numerical columns are pretty clean - Balance, Tenure, and EstimatedSalary have zero outliers, so we don't need to worry about weird values messing things up there.

Age is the biggest problem - 318 outliers is a lot. Probably elderly customers or maybe data entry errors? Either way, we should look into this before training because age seems important for churn.

CreditScore has 13 outliers, NumOfProducts has 50. Not massive numbers but worth keeping an eye on. The product one makes sense though - someone having zero products or like 4+ is unusual.

Overall the data quality is decent since most features don't have crazy outliers. Just need to decide how to handle those age outliers

### ML Pipeline

In this section the students are required to create a ML pipeline to predict whether the customer left the bank or not.

In [41]:
# import packages
import pandas as pd
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import numpy as np

In [42]:
# Load the dataset
df = pd.read_csv('Churn_Modelling_train_test.csv')


In [43]:
class Transformer:
    def __init__(self):
        self.DROP_COLUMNS = ['CustomerId', 'Surname']
        self.CATEGORICAL_COLS = ['Geography', 'Gender']

        self.BINARY_FEATURES = ['HasCrCard', 'IsActiveMember']
    
    def transform(self, df: pd.DataFrame) -> pd.DataFrame:
        df = df.drop(columns=self.DROP_COLUMNS, errors='ignore')
        df = self._ensure_binary_int(df)

        df = pd.get_dummies(df, columns=self.CATEGORICAL_COLS, drop_first=False)
        return df
    
    def _ensure_binary_int(self, df: pd.DataFrame) -> pd.DataFrame:
        for col in self.BINARY_FEATURES:
            if col in df.columns:
                df[col] = df[col].astype(int)
        return df

In [44]:

from sklearn.model_selection import train_test_split

X = df.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42,
    stratify=y  
)

train_df = X_train.copy()
train_df['Exited'] = y_train

In [45]:
from sklearn.model_selection import train_test_split

X = df.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.2, 
    random_state=42,
    stratify=y  
)

train_df = X_train.copy()
train_df['Exited'] = y_train

def balance_dataset(df: pd.DataFrame) -> pd.DataFrame:
    df_y0 = df[df['Exited'] == 0]   
    df_y1 = df[df['Exited'] == 1]  
    min_size = len(df_y1)
    df_y0_balanced = df_y0.sample(n=min_size, random_state=42)
    df_balanced = pd.concat([df_y0_balanced, df_y1])
    df_balanced = df_balanced.sample(frac=1, random_state=42).reset_index(drop=True)
    return df_balanced

train_balanced = balance_dataset(train_df)

X_train_bal = train_balanced.drop('Exited', axis=1)
y_train_bal = train_balanced['Exited']

print(f"Train original: {len(train_df)}")
print(f"Train balanceado: {len(train_balanced)}")
print(f"Test sin balancear: {len(X_test)}")

Train original: 7200
Train balanceado: 2944
Test sin balancear: 1801


In [46]:
transformer = Transformer()

X_train_transformed = transformer.transform(X_train_bal)
X_test_transformed = transformer.transform(X_test)

X_test_transformed = X_test_transformed.reindex(columns=X_train_transformed.columns, fill_value=0)

medians = X_train_transformed.median()
X_train_transformed = X_train_transformed.fillna(medians)
X_test_transformed = X_test_transformed.fillna(medians)

print("Shape X_train:", X_train_transformed.shape)
print("Shape X_test:", X_test_transformed.shape)
print("NaN in train:", X_train_transformed.isnull().sum().sum())
print("NaN in test:", X_test_transformed.isnull().sum().sum())

Shape X_train: (2944, 14)
Shape X_test: (1801, 14)
NaN in train: 0
NaN in test: 0


/var/folders/m3/gtxgx72s3qb0bls8dd96lsy40000gn/T/ipykernel_11422/1993022381.py:18: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df[col] = df[col].astype(int)
/var/folders/m3/gtxgx72s3qb0bls8dd96lsy40000gn/T/ipykernel_11422/1993022381.py:18:

In [47]:
# Create and train the model
model = LogisticRegression(max_iter=1000, random_state=42)
model.fit(X_train_transformed, y_train_bal)

print("Model trained successfully")

Model trained successfully


/Users/mac/Desktop/Eada Masters/Semester 3/ML operations and System Design /Excercises /mlops-and-system-design/.venv/lib/python3.14/site-packages/sklearn/linear_model/_logistic.py:406: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [48]:
# Predict on the test set
y_pred = model.predict(X_test_transformed)
print("First 10 predictions:", y_pred[:10])

First 10 predictions: [1 0 0 1 0 0 1 0 1 1]


In [49]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy on test: {accuracy:.4f}")
print(f"F1-score on test: {f1:.4f}")

Accuracy on test: 0.6991
F1-score on test: 0.4848


### MLFlow Integration

In this section the students are required to track the model and the parameters into MLFlow.

In [50]:
# import packages for mlfow
import mlflow

Start a local Tracking server on port 8080 - it is necessary to import mlflow and open a new terminal

In [51]:
# Set our tracking server uri for logging on port 8080
mlflow.set_tracking_uri("http://127.0.0.1:8080")

Visit the MLFlow UI on `http://127.0.0.1:8080`

In [52]:
# Create a new MLflow Experiment called `Practice Experiment - {Your Name}`

mlflow.set_experiment("Practice Experiment - Syed")

2026/05/25 19:39:01 INFO mlflow.tracking.fluent: Experiment with name 'Practice Experiment - Syed' does not exist. Creating a new experiment.


<Experiment: artifact_location='mlflow-artifacts:/2', creation_time=1779723541545, experiment_id='2', last_update_time=1779723541545, lifecycle_stage='active', name='Practice Experiment - Syed', tags={}, trace_location=None, workspace='default'>

In [53]:
# Define hyperparameters used
params = {
    "model_type": "LogisticRegression",
    "max_iter": 1000,
    "random_state": 42,
    "balanced_train": True,
    "imputation": "median",
}

# Start an MLflow run
with mlflow.start_run():
    mlflow.log_params(params)
    mlflow.log_metric("accuracy", accuracy)
    mlflow.log_metric("f1_score", f1)
    mlflow.set_tag("Training Info", "Logistic Regression on Churn dataset (balanced train)")
    mlflow.sklearn.log_model(model, "model")
    print("MLflow run logged successfully WITH model artifact")

2026/05/25 19:39:01 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/05/25 19:39:01 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


MLflow run logged successfully WITH model artifact
🏃 View run burly-bear-351 at: http://127.0.0.1:8080/#/experiments/2/runs/4ca2ef9dc2da45c9bda476a7ed6e9ea8
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/2


### Keep practicing

In [54]:
# Create a new experiment using other models or different parameters

from sklearn.ensemble import RandomForestClassifier

rf_params = {
    "model_type": "RandomForestClassifier",
    "n_estimators": 200,
    "max_depth": 10,
    "random_state": 42,
}

rf_model = RandomForestClassifier(
    n_estimators=rf_params["n_estimators"],
    max_depth=rf_params["max_depth"],
    random_state=rf_params["random_state"],
)
rf_model.fit(X_train_transformed, y_train_bal)

y_pred_rf = rf_model.predict(X_test_transformed)
accuracy_rf = accuracy_score(y_test, y_pred_rf)
f1_rf = f1_score(y_test, y_pred_rf)

print(f"Random Forest - Accuracy: {accuracy_rf:.4f}")
print(f"Random Forest - F1-score: {f1_rf:.4f}")

with mlflow.start_run():
    mlflow.log_params(rf_params)
    mlflow.log_metric("accuracy", accuracy_rf)
    mlflow.log_metric("f1_score", f1_rf)
    mlflow.set_tag("Training Info", "Random Forest on Churn dataset (balanced train)")
    mlflow.sklearn.log_model(rf_model, "model")
    print("Random Forest run logged WITH model artifact")

2026/05/25 19:39:04 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/05/25 19:39:04 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


Random Forest - Accuracy: 0.7823
Random Forest - F1-score: 0.5830
Random Forest run logged WITH model artifact
🏃 View run sneaky-wolf-129 at: http://127.0.0.1:8080/#/experiments/2/runs/7b800bdd7d3b499eb0680e7648bc5178
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/2


In [55]:
import mlflow
print("Tracking URI:", mlflow.get_tracking_uri())
print("All experiments:", mlflow.search_experiments())


Tracking URI: http://127.0.0.1:8080
All experiments: [<Experiment: artifact_location='mlflow-artifacts:/2', creation_time=1779723541545, experiment_id='2', last_update_time=1779723541545, lifecycle_stage='active', name='Practice Experiment - Syed', tags={}, trace_location=None, workspace='default'>, <Experiment: artifact_location='mlflow-artifacts:/1', creation_time=1779218110386, experiment_id='1', last_update_time=1779218110386, lifecycle_stage='active', name='Churn Prediction - Exercise', tags={}, trace_location=None, workspace='default'>, <Experiment: artifact_location='mlflow-artifacts:/0', creation_time=1779218015193, experiment_id='0', last_update_time=1779218015193, lifecycle_stage='active', name='Default', tags={}, trace_location=None, workspace='default'>]


In [56]:
import mlflow

mlflow.set_tracking_uri("http://127.0.0.1:8080")
mlflow.set_experiment("Practice Experiment - Syed")
print("Experiment created on server")

Experiment created on server


In [57]:
# Your metrics from earlier
accuracy = 0.6991  # from your cell 22
f1 = 0.4848  # from your cell 22

params = {
    "model_type": "LogisticRegression",
    "max_iter": 1000,
    "random_state": 42,
    "balanced_train": True,
    "imputation": "median",
}

with mlflow.start_run():
    mlflow.log_params(params)
    mlflow.log_metric("accuracy", accuracy)
    mlflow.log_metric("f1_score", f1)
    mlflow.set_tag("Training Info", "Logistic Regression on Churn dataset (balanced train)")
    print("Logistic Regression run logged to server")

Logistic Regression run logged to server
🏃 View run gaudy-penguin-983 at: http://127.0.0.1:8080/#/experiments/2/runs/63ad4ab3f3f1426092d5487ceee76989
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/2


In [58]:
accuracy_rf = 0.7823  # from your cell 29
f1_rf = 0.5830  # from your cell 29

rf_params = {
    "model_type": "RandomForestClassifier",
    "n_estimators": 200,
    "max_depth": 10,
    "random_state": 42,
}

with mlflow.start_run():
    mlflow.log_params(rf_params)
    mlflow.log_metric("accuracy", accuracy_rf)
    mlflow.log_metric("f1_score", f1_rf)
    mlflow.set_tag("Training Info", "Random Forest on Churn dataset (balanced train)")
    print("Random Forest run logged to server")

Random Forest run logged to server
🏃 View run flawless-conch-442 at: http://127.0.0.1:8080/#/experiments/2/runs/118282c9481e48a482ad68e6ae4316a1
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/2


In [ ]:
experiments = mlflow.search_experiments()
for exp in experiments:
    print(f"Name: {exp.name}, ID: {exp.experiment_id}")

Name: Practice Experiment - Syed, ID: 2
Name: Churn Prediction - Exercise, ID: 1
Name: Default, ID: 0


: 